syntax = "proto3";

package biometry;

service Gatekeeper {
  // --- Пользователи ---
  rpc RegisterUser (RegisterUserRequest) returns (IdResponse);
  rpc AddBiometry (AddBiometryRequest) returns (Empty); // NEW
  rpc RemoveUser (IdRequest) returns (Empty);
  rpc ListUsers (ListUsersRequest) returns (ListUsersResponse);
  
  // --- Инфраструктура ---
  rpc AddZone (AddZoneRequest) returns (IdResponse);
  rpc ListZones (ListZonesRequest) returns (ListZonesResponse);
  
  rpc AddRoom (AddRoomRequest) returns (IdResponse);
  rpc ListRooms (ListRoomsRequest) returns (ListRoomsResponse);
  
  rpc AddDevice (AddDeviceRequest) returns (IdResponse);
  rpc ListDevices (ListDevicesRequest) returns (ListDevicesResponse);
  rpc RemoveDevice (IdRequest) returns (Empty);

  // --- Контроль Доступа ---
  rpc GrantAccess (GrantAccessRequest) returns (Empty); 
  rpc SetAccessRules (SetAccessRulesRequest) returns (Empty);
  rpc GetUserAccess (IdRequest) returns (GetUserAccessResponse);

  rpc CheckAccess (CheckAccessRequest) returns (AccessCheckResponse);
  rpc CheckAccessV2 (CheckAccessRequestV2) returns (AccessCheckResponseV2);

  // --- Управление системой и Оборудованием ---
  rpc GetSystemStatus (Empty) returns (SystemStatusResponse); 
  rpc ControlService (ControlServiceRequest) returns (StatusResponse); 
  rpc ControlDoor (ControlDoorRequest) returns (StatusResponse); 
  rpc ScanHardware (Empty) returns (ScanHardwareResponse); 
  rpc ApplyRuntimeMode (RuntimeModeRequest) returns (RuntimeModeResponse);
  
  // --- Журнал ---
  rpc GetLogs (GetLogsRequest) returns (GetLogsResponse);
  rpc Shutdown(Empty) returns (Empty);
}


service DoorAgent {
  rpc SubmitObservation (DoorObservationRequest) returns (DoorObservationResponse);
  rpc GetDoorAgentStatus (Empty) returns (DoorAgentStatusResponse);
}

service Vision {
  rpc ProcessFace (ImageFrame) returns (BioResult);
  rpc GetStatus (Empty) returns (ServiceStatus);
  rpc Shutdown(Empty) returns (Empty);
}

service Audio {
  rpc ProcessVoice (AudioChunk) returns (BioResult);
  rpc GetStatus (Empty) returns (ServiceStatus);
  rpc Shutdown(Empty) returns (Empty);
}

// --- Сообщения ---

message Empty {}

message IdRequest {
  int32 id = 1;
}

message IdResponse {
  int32 id = 1;
}

// Users
message RegisterUserRequest {
  string name = 1;
  repeated bytes images = 2; // CHANGED to repeated
  repeated bytes voices = 3; // NEW
}

message AddBiometryRequest {
  int32 user_id = 1;
  repeated bytes images = 2;
  repeated bytes voices = 3;
}

message ListUsersRequest {}

message User {
  int32 id = 1;
  string name = 2;
}

message ListUsersResponse {
  repeated User users = 1;
}

// Infrastructure
message AddZoneRequest {
  string name = 1;
}

message Zone {
  int32 id = 1;
  string name = 2;
}

message ListZonesRequest {}
message ListZonesResponse {
  repeated Zone zones = 1;
}

message AddRoomRequest {
  string name = 1;
  int32 zone_id = 2;
}

message Room {
  int32 id = 1;
  int32 zone_id = 2;
  string name = 3;
}

message ListRoomsRequest {}
message ListRoomsResponse {
  repeated Room rooms = 1;
}

message AddDeviceRequest {
  string name = 1;
  int32 room_id = 2;
  string device_type = 3; 
  string connection_string = 4; 
}

message Device {
  int32 id = 1;
  int32 room_id = 2;
  string name = 3;
  string device_type = 4;
  string connection_string = 5;
  bool is_active = 6;
}

message ListDevicesRequest {}
message ListDevicesResponse {
  repeated Device devices = 1;
}

// Access
message GrantAccessRequest {
  int32 user_id = 1;
  int32 zone_id = 2;
}

message SetAccessRulesRequest {
  int32 user_id = 1;
  repeated int32 allowed_room_ids = 2;
  repeated int32 allowed_zone_ids = 3;
}

message GetUserAccessResponse {
  repeated int32 allowed_room_ids = 1;
  repeated int32 allowed_zone_ids = 2;
}

message CheckAccessRequest {
  int32 device_id = 1;
  bytes image = 2;
  bytes audio = 3;
  uint32 audio_sample_rate = 4;
}

message AccessCheckResponse {
  bool granted = 1;
  string user_name = 2;
  string message = 3;
  bool face_detected = 4;
  bool face_live = 5;
  float face_liveness_score = 6;
  float face_distance = 7;
  bool face_match = 8;
  bool voice_provided = 9;
  bool voice_detected = 10;
  float voice_distance = 11;
  bool voice_match = 12;
  float final_confidence = 13;
  string decision_stage = 14;
}

enum AccessStage {
  ACCESS_STAGE_UNKNOWN = 0;
  ACCESS_STAGE_PRESENCE = 1;
  ACCESS_STAGE_FACE_QUALITY = 2;
  ACCESS_STAGE_FACE_LIVENESS = 3;
  ACCESS_STAGE_FACE_MATCH = 4;
  ACCESS_STAGE_VOICE_OPTIONAL = 5;
  ACCESS_STAGE_POLICY = 6;
  ACCESS_STAGE_GRANTED = 7;
  ACCESS_STAGE_DENIED = 8;
}

message CheckAccessRequestV2 {
  string session_id = 1;
  int32 device_id = 2;
  repeated bytes frames = 3;
  bytes audio = 4;
  uint32 audio_sample_rate = 5;
  repeated int64 frame_timestamps_ms = 6;
}

message AccessCheckResponseV2 {
  bool granted = 1;
  AccessStage stage = 2;
  string reason = 3;
  float confidence = 4;
  int32 user_id = 5;
  string user_name = 6;
  bool face_detected = 7;
  float face_quality = 8;
  float face_live_score = 9;
  bool face_live = 10;
  float face_distance = 11;
  bool face_match = 12;
  bool voice_provided = 13;
  bool speech_present = 14;
  float voice_live_score = 15;
  bool voice_live = 16;
  float voice_distance = 17;
  bool voice_match = 18;
  bool suspicious_audio = 19;
  repeated string flags = 20;
  string session_id = 21;
}

// Vision & Audio
message ImageFrame {
  bytes content = 1;
}

message AudioChunk {
  bytes content = 1;
  uint32 sample_rate = 2;
}

message BioResult {
  bool detected = 1;
  bool is_live = 2;
  float liveness_score = 3;
  repeated float embedding = 4;
  string error_msg = 5;
  string execution_provider = 6;
}



enum DoorPipelineStage {
  DOOR_STAGE_IDLE = 0;
  DOOR_STAGE_PRESENCE = 1;
  DOOR_STAGE_COLLECTING_CLIP = 2;
  DOOR_STAGE_IDENTIFICATION = 3;
  DOOR_STAGE_COOLDOWN = 4;
}

message DoorObservationRequest {
  string session_id = 1;
  int32 device_id = 2;
  bytes frame = 3;
  int64 timestamp_ms = 4;
  bytes audio = 5;
  uint32 audio_sample_rate = 6;
}

message DoorObservationResponse {
  bool access_granted = 1;
  DoorPipelineStage stage = 2;
  string reason = 3;
  float confidence = 4;
  string user_name = 5;
  bool pending = 6;
  repeated string flags = 7;
}

message DoorAgentStatusResponse {
  int32 active_sessions = 1;
  int32 cooldown_sessions = 2;
}

// System Management
message ServiceStatus {
  bool online = 1;
  string device = 2; // "CUDA", "CPU"
  string message = 3; // "OK" or error details
}

message SystemStatusResponse {
  ServiceStatus vision = 1;
  ServiceStatus audio = 2;
  ServiceStatus database = 3;
  ServiceStatus gateway = 4;
}

message ControlServiceRequest {
  string service_name = 1; 
  string action = 2; 
}

message StatusResponse {
  bool success = 1;
  string message = 2;
}

message ControlDoorRequest {
  int32 device_id = 1;
  string command = 2; 
}

message ScanHardwareResponse {
  repeated Device found_devices = 1; 
}

message RuntimeModeRequest {
  string mode = 1; // auto | cpu | gpu
  bool restart_services = 2;
}

message RuntimeModeResponse {
  bool success = 1;
  string message = 2;
  string saved_mode = 3;
  int32 cpu_cores = 4;
  int32 cpu_threads = 5;
  bool gpu_available = 6;
  int32 vision_threads = 7;
  int32 audio_threads = 8;
}

// Logs
message GetLogsRequest {
  int32 limit = 1;
  int32 offset = 2;
}

message LogEntry {
  int32 id = 1;
  string timestamp = 2;
  string user_name = 3;
  string room_name = 4;
  string zone_name = 5;
  bool access_granted = 6;
  string details = 7;
}

message GetLogsResponse {
  repeated LogEntry logs = 1;
}
